win32ss/printing/base/winspool/utils.c

```diff
@@ -0,0 +1,62 @@
+/*
+* PROJECT:     ReactOS Spooler API
+* LICENSE:     GPL-2.0+ (https://spdx.org/licenses/GPL-2.0+)
+* PURPOSE:     Utility Functions related to Print Processors
+* COPYRIGHT:   Copyright 2020 Doug Lyons (douglyons@douglyons.com)
+*/
+
+#include <strsafe.h>
+
+BOOL UnicodeToAnsiInPlace(PWSTR pwszField)
+{
+    /*
+     * This converts an incoming Unicode string to an ANSI string.
+     * It returns FALSE on failure, otherwise it returns TRUE.
+     * It is only useful for "in-place" conversions where the ANSI string goes
+     * back into the same place where the Unicode string came into this function.
+     * It seems that many of the functions involving printing can use this.
+     */
+
+    PSTR pszField;
+    DWORD cch;
+    BOOL bReturn = TRUE;
+
+    /*
+     * Map the incoming Unicode pwszField string to an ANSI one here so that we can do
+     * in-place conversion. We read the Unicode input and then we write back the ANSI
+     * conversion into the same buffer for use with our GetPrinterDriverA function
+     */
+    PSTR pwszOutput = (PSTR)pwszField;
+
+    if (!pwszField)
+    {
+        goto Exit;
+    }
+    else
+    {
+        cch = wcslen(pwszField);
+    }
+
+    if (cch == 0)
+    {
+        goto Exit;
+    }
+
+    pszField = HeapAlloc(hProcessHeap, 0, (cch + 1) * sizeof(CHAR));
+    if (!pszField)
+    {
+        SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+        ERR("HeapAlloc failed!\n");
+        bReturn = FALSE;   // indicates a failure to be handled by caller
+        goto Exit;
+    }
+
+    WideCharToMultiByte(CP_ACP, 0, pwszField, -1, pszField, cch + 1, NULL, NULL);
+    StringCchCopyA(pwszOutput, cch + 1, pszField);
+
+    HeapFree(hProcessHeap, 0, pszField);
+
+Exit:
```

Comment:
Cleanup seemed to indicate that you were doing more than just Exiting such as cleaning up variables and I was only trying to create a single exit point to Exit. Please see if you are OK with these changes. Thanks.

Reply:
@Doug-Lyons: Just my opinion, but this is too much verbose, it is a source not a readme, isn't better something like this (for example)?

```
case 6:
{
	if( !UnicodeToAnsiInPlace(pdi6w->pszMfgName) || !UnicodeToAnsiInPlace(pdi6w->pszOEMUrl) ... )
		return FALSE;
}

return TRUE;
```

A single exit point isn't an advantage if there isn't common code at the end but instead it is a disadvantage (one additional "goto" instruction to execute without advantages).

Reply:
@ale5000-git, I do not disagree, but this comment was connected to a line in utils.c and you are referencing printers.c. Since @ColinFinck is the code owner, I would like to have his opinion on what he would like to see here and I will be happy to make any changes that he suggests. Thanks.

Reply:
For the printers.c code @ale5000-git is talking about, I would have also gone for `return FALSE` or `return TRUE` instead of a `goto`. But I don't have strong opinions on that one. @Doug-Lyons final code in this PR is also fine for me. The compiler should optimize both variants to the same instructions.

Merging this PR now